primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,163 @@
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+}
+
+/// A single executor parameter
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutorParam {
+	/// ## General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
+
+	/// ## Parameters setting the executuion environment semantics:
+	/// Number of extra heap pages
+	ExtraHeapPages(u64),
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit, in units
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+	/// Executor instantiation strategy
+	InstantiationStrategy(ExecInstantiationStrategy),
+	/// `true` if compiler should perform NaNs canonicalization
+	CanonicalizeNaNs(bool),
+	/// `true` if parallel compilation is allowed, single thread is used otherwise
+	ParallelCompilation(bool),
+}
+
+impl ExecutorParam {
+	fn is_environment(&self) -> bool {
+		match self {
+			ExecutorParam::Environment(_) => true,
+			_ => false,
+		}
+	}
+}
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl ExecutorParamsHash {
+	/// Create a new executor parameter hash from `H256` hash
+	pub fn from_hash(hash: Hash) -> Self {
+		Self(hash)
+	}
+}
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+/// # Deterministically serialized execution environment semantics
+/// Represents an arbitrary semantics of an arbitrary execution environment, so should be kept as
+/// abstract as possible.
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub struct ExecutorParams(Vec<ExecutorParam>);
+
+impl ExecutorParams {
+	/// Creates a new, empty executor parameter set
+	pub fn new(environment: ExecutionEnvironment) -> Self {
+		ExecutorParams(vec![ExecutorParam::Environment(environment)])
+	}
+
+	/// Returns execution environment type
+	pub fn environment(&self) -> ExecutionEnvironment {
+		debug_assert!(self.0.len() > 0);
+		if let ExecutorParam::Environment(environment) = self.0[0] {
```

Comment:
If this field is mandatory, why put it in the `Vec`? and not create a separate struct member for it?

Reply:
Well, as for me, it's more about long-term design/philosophy than about an idiomatic implementation. For example, in this very PR, we're struggling to update a base primitive, `SessionInfo`. I've already tormented at least three persons on how to do it the best way, and still no clear pathway. `ExecutorParams` is much more volatile; we don't know what comes and what goes.

My current design is migration-proof. You can put whatever you want into `ExecutorParams` and stop using whatever you don't want to use at any moment.

Now, if we split fields into "mandatory" and "optional" ones, what if we add another mandatory one? Either migrate the structure or put a new mandatory param inside the Vec. I bet anyone doing it would put it inside the Vec to keep things simple. And then we have a terrible design with mandatory params partially outside of Vec and partially inside it.

So I decided to keep it homogeneous and to enforce mandatoriness in-code. But, again, feel free to make me change my mind.

Reply:
This is a good argument and convinced me - please add something like that to the code.